pkg/nsx/services/securitypolicy/firewall.go

```diff
@@ -706,26 +710,64 @@ func (service *SecurityPolicyService) deleteSecurityPolicy(sp types.UID) error {
 		return err
 	}
 
-	err = securityPolicyStore.Apply(&finalSecurityPolicyCopy)
+	// Apply security policy and rules to store
+	err = service.applySecurityPolicyStore(finalSecurityPolicyCopy, finalSecurityPolicyCopy.Rules, *nsxGroupsEmpty, true)
 	if err != nil {
-		log.Error(err, "Failed to apply store", "securityPolicy", finalSecurityPolicyCopy)
 		return err
 	}
-	err = ruleStore.Apply(&finalSecurityPolicyCopy.Rules)
-	if err != nil {
-		log.Error(err, "Failed to apply store", "nsxRules", finalSecurityPolicyCopy.Rules)
-		return err
-	}
-	err = groupStore.Apply(nsxGroups)
+
+	// Delete groups after security policy and rules are deleted
+	err = service.deleteGroupsWithStore(nsxGroups, nil, groupStore)
 	if err != nil {
-		log.Error(err, "Failed to apply store", "nsxGroups", nsxGroups)
 		return err
 	}
 
 	log.Info("Successfully deleted NSX SecurityPolicy", "nsxSecurityPolicy", finalSecurityPolicyCopy)
 	return nil
 }
 
+// deleteGroupsWithStore deletes NSX groups from NSX and the group store.
+// It handles ResourceInUse errors by logging them but continuing execution.
+// Returns an error if deletion fails with a non-ResourceInUse error or if applying to the store fails.
+func (service *SecurityPolicyService) deleteGroupsWithStore(nsxGroups *[]model.Group, vpcInfo *common.VPCResourceInfo, groupStore *GroupStore) error {
+	if len(*nsxGroups) == 0 {
+		return nil
+	}
+
+	for _, group := range *nsxGroups {
+		groupID := *group.Id
+		var err error
+
+		// Use the appropriate client based on whether VPC is enabled
+		if IsVPCEnabled(service) {
+			err = service.NSXClient.VpcGroupClient.Delete(vpcInfo.OrgID, vpcInfo.ProjectID, vpcInfo.VPCID, groupID)
+		} else {
+			failIfSubtreeExists, forceDelete := true, true
```

Comment:
Why we set these two values to true? I have the question because the default values are false. And I suspect that if the forceDelete is true, we cannot capture the IN_USE error.

Reply:
OK, better use the default.